java/ql/src/experimental/Security/CWE/CWE-094/SpringViewManipulationLib.qll

```diff
@@ -0,0 +1,39 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import SpringFrameworkLib
+
+/**
+ * A taint-tracking configuration for unsafe user input
+ * that can lead to Spring View Manipulation vulnerabilities.
+ */
+class SpringViewManipulationConfig extends TaintTracking::Configuration {
+  SpringViewManipulationConfig() { this = "Spring View Manipulation Config" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof RemoteFlowSource or
+    source instanceof WebRequestSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof SpringViewManipulationSink }
+}
+
+/**
+ * A sink for Spring View Manipulation vulnerabilities,
+ */
+class SpringViewManipulationSink extends DataFlow::ExprNode {
+  SpringViewManipulationSink() {
+    exists(ReturnStmt r, SpringRequestMappingMethod m |
+      r.getResult() = this.asExpr() and
+      m.getBody().getAStmt() = r and
+      not m.isResponseBody() and
+      r.getResult().getType() instanceof TypeString
```

Comment:
We observed many False Positives caused by controller methods performing redirects such as:

```
 return "redirect:" + request.getHeader("Referer");
```

 or

```
 return redirectToRenderSession(bar, foo);
```

These are open for a new separate bounty submission for Open Redirect, but not relevant in this context. Note that `redirect:` prefix can be added directly in the return statement or in a method such as the second example, so the sanitizer should look for string concatenations/interpolations in any node of the dataflow

Reply:
okay I will extract this and create a new PR for the Open Redirect issue.